src/sudoku.rs

```diff
@@ -0,0 +1,223 @@
+// http://rosettacode.org/wiki/Sudoku
+
+#![allow(unstable)]
+
+use std::{fmt, iter};
+use std::str::FromStr;
+use std::num::Int;
+
+const BOARD_WIDTH: usize = 9;
+const BOARD_HEIGHT: usize = 9;
+const GROUP_WIDTH: usize = 3;
+const GROUP_HEIGHT: usize = 3;
+const MAX_NUMBER: usize = 9;
+
+type BITS = u16;
+const MASK_ALL: BITS = 0x1ff;
+const INVALID_CELL: usize = !0;
+
+#[derive(Copy, Eq, PartialEq)]
+struct Sudoku {
+    map: [[BITS; BOARD_WIDTH]; BOARD_HEIGHT]
+}
+
+impl Sudoku {
+    fn new() -> Sudoku {
+        Sudoku { map: [[MASK_ALL; BOARD_WIDTH]; BOARD_HEIGHT] }
+    }
+
+    fn get(&self, x: usize, y: usize) -> usize {
+        match self.map[y][x].count_ones() {
+            0 => INVALID_CELL,
+            1 => self.map[y][x].trailing_zeros() + 1,
+            _ => 0
+        }
+    }
+
+    fn set(&mut self, x: usize, y: usize, n: usize) {
+        self.map[y][x] = 1 << (n - 1);
+    }
+
+}
+
+impl FromStr for Sudoku {
+    fn from_str(s: &str) -> Option<Sudoku> {
+        let mut sudoku = Sudoku::new();
+
+        for (y, line) in s.lines().filter(|l| !l.is_empty()).enumerate() {
+            let line = line.trim_matches(CharExt::is_whitespace);
+            for (x, c) in line.chars().enumerate() {
+                if let Some(d) = c.to_digit(10) {
+                    if d != 0 { sudoku.set(x, y, d); }
+                } else {
+                    return None
+                }
+            }
+        }
+
+        Some(sudoku)
+    }
+}
+
+impl fmt::String for Sudoku {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let hbar = "+---+---+---+";
+
+        for y in (0 .. BOARD_HEIGHT) {
+            if y % GROUP_HEIGHT == 0 {
+                try!(writeln!(f, "{}", hbar));
+            }
+
+            for x in (0 .. BOARD_WIDTH) {
+                if x % GROUP_WIDTH == 0 {
+                    try!(write!(f, "|"));
+                }
+
+                match self.get(x, y) {
+                    INVALID_CELL => try!(write!(f, "!")),
+                    0 => try!(write!(f, " ")),
+                    d => try!(write!(f, "{}", d))
+                }
+            }
+            try!(writeln!(f, "|"));
+        }
+        try!(writeln!(f, "{}", hbar));
+
+        Ok(())
+    }
+}
+
+fn solve_sudoku(mut puzzle: Sudoku) -> Vec<Sudoku> {
```

Comment:
Yes, `solve_sudoku` may return multiple solutions if problem is not good :smile: